### PR TITLE
[WIP] initial commit for syncing Sia in 73 minutes post

### PR DESCRIPTION
**Pull request by tbenz9**
Here is my next article if you are still interested in having me contribute to space duck.  I am still waiting for a link from RBZL for his bootstrapping the blockchain instructions.  Please feel free to proofread and offer any suggestions you have.

I wasn't sure how to introduce my code snippets, and I don't have any images which typically break up the posts.  If you have any ideas on either of those I would be interested to hear them.

Per your approval, I was targeting a Monday publish date.

### CODE REVIEW

File: _posts/2018-04-09-synching-sia.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+---
+title: Syncing Sia in 73 minutes
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: What do you think of the title "Overclocking the Sia Blockchain Sync"?

I think of this as similar to overclocking a CPU or GPU in that it's something anyone can do if they know what they're doing, but it introduces too much instability for the casual user.

I'd prefer to focus less on the "73 minutes" metric. It's impressive to a reader who knows that syncing normally takes like 14 hours, but is not so impressive to a reader coming in without that context or to a reader who thinks bootstrapping is a sensible way to sync.

I think the more impressive metric is the speedup percentage. This is a 76% reduction in sync time, right? To me, that's the more exciting metric.

File: _posts/2018-04-09-synching-sia.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+tags:
+- sia
+- blockchain
+- benchmark
+permalink: "/sia-sync-73-min/"
+---
+
+*This is a guest post by Sia community contributor [Thomas Bennett](https://github.com/tbenz9).*
+
+I can sync Sia from a block height of 0 to 149000 in 73 minutes.  Here’s how I do it.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: I feel like the intro is a bit abrupt. Can we ease the reader into it a bit more by explaining the context? We shouldn't assume they've read the previous blog post.

Why should the reader be interested and what's going to prevent them from immediately clicking the back button?

Suggest talking about:

* Motivations for why someone would want to do this
* Tell them you're going to teach them how to use this to speed up their own sync.

File: _posts/2018-04-09-synching-sia.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+## Zero subscribers
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: Can we start with an overview of the steps involved before we dive into each one? Just like a bulleted list saying each thing we changed.

File: _posts/2018-04-09-synching-sia.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+As mentioned in my previous article [Benchmarking Sia’s Blockchain Sync](https://blog.spaceduck.io/sia-blockchain-sync/) Sia spends an average of 55 minutes or 19.1% of its time notifying subscribers.  I submitted the pull request [#2809](https://github.com/NebulousLabs/Sia/pull/2809) so Sia only notifies subscribers if there are any.  By running Sia with only the `consensus` and `gateway` modules loaded there will be zero subscribers.
+
+My `siad` command with minimal modules loaded.
+
+```shell
+./siad -M cg -d /root/Sia/
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: I prefer to always use long flag names in blog post so it's clear to the reader what the flag is, so I think `--modules cg` is better.

Suggest dropping the `-d /root/Sia` part because it's not related to eliminating subscribers.

File: _posts/2018-04-09-synching-sia.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+permalink: "/sia-sync-73-min/"
+---
+
+*This is a guest post by Sia community contributor [Thomas Bennett](https://github.com/tbenz9).*
+
+I can sync Sia from a block height of 0 to 149000 in 73 minutes.  Here’s how I do it.
+
+## Zero subscribers
+
+As mentioned in my previous article [Benchmarking Sia’s Blockchain Sync](https://blog.spaceduck.io/sia-blockchain-sync/) Sia spends an average of 55 minutes or 19.1% of its time notifying subscribers.  I submitted the pull request [#2809](https://github.com/NebulousLabs/Sia/pull/2809) so Sia only notifies subscribers if there are any.  By running Sia with only the `consensus` and `gateway` modules loaded there will be zero subscribers.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: Suggest rephrasing the mention of the PR to make it clear that this functionality is already available in 1.3.2. I had to check the PR merge date against the 1.3.2 release date to figure it out.

File: _posts/2018-04-09-synching-sia.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+```shell
+./siad -M cg -d /root/Sia/
+```
+
+With zero subscribers Sia skips the notify subscribers functions and shaves 55 minutes off the synchronization time.  Note that with only the `consensus` and `gateway` modules loaded you cannot rent, host, or use the Sia wallet.
+
+## Reduce RPC Timeouts
+
+Sia has long timeouts when waiting for blocks to download.  I’ve submitted the pull request [#2922](https://github.com/NebulousLabs/Sia/pull/2922) to reduce some of the long timeouts.  In order to sync in 73 minutes, I dropped the timeouts in [syncronize.go](https://github.com/NebulousLabs/Sia/blob/master/modules/consensus/synchronize.go) **WAY** down.  I first tried setting `relayHeaderTimeout`, `sendBlkTimeout` and `sendBlocksTimeout` to 2 seconds, this was too low, I was timing out on all my peers and couldn’t download any blocks.  After bumping `sendBlkTimeout` and `sendBlocksTimeout` to 10 seconds I was able to stay connected to most peers and download the blocks very quickly.  Any peers that could not deliver my batch of 10 blocks in 10 seconds or less was disconnected and Sia would look for new peers.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: >I first tried setting `relayHeaderTimeout`, `sendBlkTimeout` and `sendBlocksTimeout` to 2 seconds

We need to introduce these first. Most readers aren't going to be familiar enough with the Sia codebase to know what these functions are. They should be able to understand the blog post without clicking to any external links.

If we can explain these timeouts in prose rather than by variable name, that would be ideal. Like in plain English, what are these events Sia is waiting on?

Also, in general I recommend linking by commit ID rather than branch. Commit IDs stay forever, whereas if they rewrite the file or rename it, a link to the branch will break:

https://github.com/NebulousLabs/Sia/blob/49704703740e12aa59abda5606346faac411f0b4/modules/consensus/synchronize.go

File: _posts/2018-04-09-synching-sia.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+My 73-minute sync only spent 168 seconds downloading blocks from peers.  This number is actually slightly confusing to me since if Sia downloaded 10GB of blocks it exceeded my advertised home internet speed.  A couple of things could have happened here:
+The downloaded blocks could be less than 10GB in size when sent over the Internet, I never measured how much data I actually received from my NIC.
+I could have a mistake in my timing function causing the elapsed download time to be reported wrong.
+My ISP could allow me to burst into higher speed tiers during low network congestion.
+
+Regardless of why my download speed seems off, I was able to sync from block 0 to current height in 73 minutes.
+
+The timeouts I set in `synchronize.go` for my 73-minute sync.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: Can we link to a diff of your test code against the latest commit? 

Each of the other optimizations includes instructions for how the user can replicate the results, so I think these should too. Something kind of like:

```shell
go get -u github.com/NebulousLabs/Sia/...
cd $GOPATH/src/github.com/NebulousLabs/Sia
git remote add tbenz9 https://github.com/tbenz9/Sia.git
# Not sure the exact syntax but check out the speedup code from your branch
git checkout -b reduce-timeouts tbenz9/reduce-timeouts
make release-std
$GOPATH/bin/siad ...
```

File: _posts/2018-04-09-synching-sia.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+/dev/mapper/pve-vm--200--disk--1   32G  2.1G   28G   7% /
+none                              492K     0  492K   0% /dev
+tmpfs                              32G     0   32G   0% /dev/shm
+tmpfs                              32G  8.2M   32G   1% /run
+tmpfs                             5.0M     0  5.0M   0% /run/lock
+tmpfs                              32G     0   32G   0% /sys/fs/cgroup
+tmpfs                              11G     0   11G   0% /root/Sia/consensus
+```
+
+A RAMDisk has very fast I/O but is not persistent across reboots.  Most users are going to want to copy the `consensus.db` file to persistent storage after the sync is done.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: I think it's safe to assume nearly all users would want to move it to persistent storage. Can we show those steps?

File: _posts/2018-04-09-synching-sia.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+2018/04/05 20:11:49.321247 accept.go:275: Height 1000 at 1522959109
+… TRUNCATED
+2018/04/05 21:23:39.602232 accept.go:275: Height 148000 at 1522963419
+2018/04/05 21:23:39.602263 accept.go:276: Download time: 168.738052889
+2018/04/05 21:23:39.602270 accept.go:277: Validate time: 2631.783831114
+2018/04/05 21:23:39.602274 accept.go:278: Apply time: 250.713886247
+2018/04/05 21:23:39.602278 accept.go:279: Subscribers time: 0.01789937
+… TRUNCATED
+2018/04/05 21:24:57.406911 synchronize.go:642: INFO: IBD done, synced with 6 peers
+```
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: Suggest cutting lines 107-120. It adds a lot of space, but I don't think it's that interesting for the reader. At most, I think we should have a "Raw data" section with simple links (like in the [load test posts](https://blog.spaceduck.io/load-test-2/#raw-data)).

File: _posts/2018-04-09-synching-sia.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+tmpfs                              32G     0   32G   0% /dev/shm
+tmpfs                              32G  8.2M   32G   1% /run
+tmpfs                             5.0M     0  5.0M   0% /run/lock
+tmpfs                              32G     0   32G   0% /sys/fs/cgroup
+tmpfs                              11G     0   11G   0% /root/Sia/consensus
+```
+
+A RAMDisk has very fast I/O but is not persistent across reboots.  Most users are going to want to copy the `consensus.db` file to persistent storage after the sync is done.
+
+## Future optimizations
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: These two feel weird paired together. The first one I feel like falls in the scope of things the user could trivially do on their own if they want to. The second is something that the devs probably have to spend months doing.

The first one fits with this article, but I think the reader might wonder why the experiment skipped this seemingly easy optimization. Can we explain that?

The second one, consider cutting because it feels outside the scope of the article.

File: _posts/2018-04-09-synching-sia.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+Sia downloads blocks in batches of 10. That means Sia requests blocks at least 14,800 times for a full blockchain sync. I believe Sia could request blocks in batches of 25, or even 100 and reduce some of the overhead from making so many requests. I proposed that the devs raise the batch number to 25 but my proposal was denied. You can read the discussion and compelling arguments from both sides in my pull request [#2922](https://github.com/NebulousLabs/Sia/pull/2922).
+
+### A consensus overhaul
+
+Many optimizations such as parallel block downloads, smarter timeouts, I/O buffering, and header-first block retrieval could be added with a consensus package overhaul.  I don’t believe this is the best use of time for the limited dev resources Nebulous currently has, but I do believe an overhaul will be necessary if Sia and its blockchain continue to grow.
+
+Concerns about the blockchain size and performance have been brought up several times, and the Sia devs have several ideas on how to support a large blockchain.  You can read some of the discussion on GitHub in issue [#2897](https://github.com/NebulousLabs/Sia/issues/2897).
+
+## A 73-minute sync is mostly academic
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: I think this section kind of undermines the article. A lot of the methods you mentioned are practical if you think of it as overclocking. Can we take this section out but move the content into the relevant sections?

Like I think the stuff about RAMDisk is relevant and should just live in that section. It avoids I/O bottlenecks but you need 11 GB of RAM free and you have to awkwardly shuffle the file around.

Same thing with the timeouts. They're practical, but you'll have to adjust them based on your connection speed. The ones you used worked for a connection of 200 Mbps.

File: _posts/2018-04-09-synching-sia.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+Note that 10-second timeouts are not very friendly to the Sia network and 10 seconds is not feasible for users with slower internet connections.  For example, the average Tor user or 3rd world Internet user would not be able to sync Sia with `sendBlocksTimeout` set to 10 seconds.  For this reason, the default `sendBlocksTimeout` in my pull request has been set at 180 seconds to allow for users with slower internet connections.  The Sia devs and I think this is a good balance of accommodating users with slower connections, while not wasting too much time waiting for unresponsive peers.
+
+My 73-minute sync only spent 168 seconds downloading blocks from peers.  This number is actually slightly confusing to me since if Sia downloaded 10GB of blocks it exceeded my advertised home internet speed.  A couple of things could have happened here:
+The downloaded blocks could be less than 10GB in size when sent over the Internet, I never measured how much data I actually received from my NIC.
+I could have a mistake in my timing function causing the elapsed download time to be reported wrong.
+My ISP could allow me to burst into higher speed tiers during low network congestion.
```
Comment (mtlynch) [inline _posts/2018-04-09-synching-sia.md]: Yeah, I'm skeptical of this result. That's equivalent to a download rate of 476 Mbps. Sia actually *can* go faster than most applications (including browser speed tests) because it's generating so many TCP connections to peers behind a variety of ISPs, so there are fewer links to congest (BitTorrent has the same property). But 476 just seems so suspiciously high.

Can we cut this paragraph entirely? The previous article was about breaking down the sync time into components, so it would be relevant in an article like that. But this article is just about the final 73 minute number, which we're confident in. It feels like a digression to talk about this suspicious sub-metric and speculate about it.

File: _posts/2018-04-09-synching-sia.md (hunk 1 above)
Comment (tbenz9) [reply _posts/2018-04-09-synching-sia.md]: I'm not a huge fan of the term "overclocking"; I associate overclocking with hardware tweaks such as increased voltage, cooling, and CPU multipliers.  What do you think about "Improving the Sia Blockchain Sync"?

File: _posts/2018-04-09-synching-sia.md (hunk 2 above)
Comment (tbenz9) [reply _posts/2018-04-09-synching-sia.md]: Sure, I noticed it was a bit abrupt also, I hate when blogs spend too much time introducing the article, but I'll add a sentence or two.

File: _posts/2018-04-09-synching-sia.md (hunk 10 above)
Comment (tbenz9) [reply _posts/2018-04-09-synching-sia.md]: I've clarified the 1st point (it requires network adoption before it would have any effect).  I still think these are related because they are both areas that the sync time could be improved, but they require dev time, testing, and network adoption before it would make any difference.

I wanted to include `a consensus overhaul` section because the devs have talked about it and have several ideas on how to improve it which are in the issue I linked.

File: _posts/2018-04-09-synching-sia.md (hunk 1 above)
Comment (tbenz9) [reply _posts/2018-04-09-synching-sia.md]: Or "Improving the performance of the Sia Blockchain Sync"

Comment (tbenz9) [reply _posts/2018-04-09-synching-sia.md]: Usually, the editor gets to decide the title so I'll let you make the final decision as you're acting as the editor in this case.
